### PR TITLE
WIP: Enhancement/use node for rpc

### PR DESCRIPTION
**Pull request by oskarth**
Having some issues with replacing WakuRelay with WakuNode in RPC call. Not quite sure I understand the error here.

The node reference is accurate, but for some reason it thinks I want WakuRelay. It might be because the reference doesn't contain the publish method, or I am missing something in my mental model of how this is supposed to work.

All experimental changes in last commit https://github.com/status-im/nim-waku/pull/142/commits/60c5f5a563d91ad6a834f55b428a5ba635226fcf compared to other PR
 
Error:

```
 WIP: Attempt at replacing WakuRelay with Wakunode

/home/oskarth/git/status-im/nim-waku/waku/node/v2/rpc/wakurpc.nim(39, 17) Error: type mismatch: got <WakuNode, string, WakuMessage>
but expected one of:
method publish(w: WakuRelay; pubSubTopic: string; message: seq[byte]): Future[int]
  first type mismatch at position: 1
  required type for w: WakuRelay
  but expression 'node' is of type: WakuNode

expression: publish(node, topic, message)
stack trace: (most recent call last)
```

## Aside on scoping/shadow issue (resolved)
As an aside, I'm seeing some weird scoping/shadowing issues when using RPC. E.g. in https://github.com/status-im/nim-waku/pull/142/commits/5c24b6aea543eba84e3d2a452e96bea7a30f0adc:

 Fix publish topic and payload script

Also change parameter name in waku relay due to weird shadowing of log
topic:

DBG 2020-09-09 12:07:54+08:00 waku_publish                               tid=8795 file=wakurpc.nim:30 topic=waku payload=@[]
DBG 2020-09-09 12:07:54+08:00 publish                                    tid=8795 file=waku_relay.nim:65 topic=WakuRelay

Above should show topic=waku but it gets topic=WakuRelay from log scope
for some reason.

Resolved by changing parameter name.

Expected would be that the most local variable, i.e. parameter, wins, or I get a compiler error. logScope macro magic?

### CODE REVIEW

**Comment by kdeme** [comment]
The problem is that you are missing the `wakunode2` import. But if you add it, other issues occur as `wakunode2` already imports `wakurpc`.  Although it strangely doesn't give the usual cyclic import warning/error. 

Anyhow, I think the application (cli in this case, with rpc, metrics etc.) logic should be separated from the "Waku Node API", similar like in v1. I've already started with the split of the logic itself in previous PR, so should be convenient now.  We can also decide to just call it Waku API by the way (and waku.nim alongside).

File: waku/node/v2/rpc/wakurpc.nim [hunk 1]
```diff
@@ -27,11 +27,20 @@ proc setupWakuRPC*(node: WakuNode, rpcsrv: RpcServer) =
...
     # TODO: Shouldn't we really be doing WakuNode publish here?
+    debug "waku_publish", topic=topic, payload=payload
     discard wakuRelay.publish(topic, payload)
     return true
     #if not result:
     #  raise newException(ValueError, "Message could not be posted")
 
+  rpcsrv.rpc("waku_publish2") do(topic: string, message: WakuMessage) -> bool:
+    # XXX also future return type
+    debug "waku_publish2", topic=topic, message=message
```
Comment (kdeme) [inline waku/node/v2/rpc/wakurpc.nim:38]: wakunode2 import missing, but can't add it as `wakurpc` is imported already in `wakunode2`

**Comment by oskarth** [comment]
> The problem is that you are missing the wakunode2 import. But if you add it, other issues occur as wakunode2 already imports wakurpc. Although it strangely doesn't give the usual cyclic import warning/error.

Gah sorry, I had two branches and the other one has the `wakunode2` import. There are still issues though

**Comment by oskarth** [comment]
> Anyhow, I think the application (cli in this case, with rpc, metrics etc.) logic should be separated from the "Waku Node API", similar like in v1. I've already started with the split of the logic itself in previous PR, so should be convenient now. We can also decide to just call it Waku API by the way (and waku.nim alongside).

Which PR? Can you elaborate on how you imagine the split looking like and why?

**Comment by kdeme** [comment]
This PR https://github.com/status-im/nim-waku/pull/117 paves the road to separate this better.
E.g. An application that wants to use the Waku Node API does not necessarily want all our wakunode cli application logic such as the RPC calls, metrics, logging, cli config, etc.

So, we should allow to only import Waku Node API calls, and nothing extra.

We can do this:
- either by creating two files. Say `waku.nim` (API calls) and `wakunode.nim` (CLI application). Or if you want to keep the node name in there, `wakunode.nim` (API calls) and `wakucli.nim` (CLI application).
- or by putting the purely application parts (including its imports) under the `when isMainModule`.

This would remove also the cyclic import which I'm guessing causing the problem you are seeing.

**Comment by oskarth** [comment]
Gotcha, so I suppose it makes more sense to do either of those suggested actions before moving over to using the Node API as started here.

**Comment by oskarth** [comment]
Closing this PR for now, see https://github.com/status-im/nim-waku/issues/143#issuecomment-689971546 for next steps.

**Pull request closed without merging**